### PR TITLE
Add `cms:page_file_link` tag to render page files

### LINKED ISSUE #796: How to use page files in v2
In CMS v1, there was a handy `page_file` tag that created an input field and rendered the file URL at the same time.

In CMS v2, I can add `{{ cms:files attachments }}` to the layout, but how do I render an attachment from a page?

I could use the URL, but that's significantly worse than CMS v1.

### PR DESCRIPTION
**Pull request by glebm**
Examples:

    {{ cms:page_file_link header }}
    {{ cms:page_file_link attachments, filename: "cat.jpg", as: image }}

Dragging page-level files onto the textarea generates these tags.

See #796 for more information.
Fixes #796

### CODE REVIEW

File: app/views/comfy/admin/cms/fragments/_form_fragment_attachments.html.haml [hunk 1]
```diff
@@ -4,9 +4,19 @@
...
+        cms_page_file_link_tag = [
+          "{{ cms:page_file_link #{fragment_id}",
+          (", filename: \"#{filename}\"" if multiple),
+          (", as: image" if attachment.image?),
+          " }}".html_safe
+        ].compact.join
+      = link_to truncated_filename, attachment,
+                data: {toggle: "page-file-popover", content: thumb},
+                class: "btn btn-light text-truncate", target: "_blank",
+                ondragstart: "event.dataTransfer.setData('text/plain', '#{cms_page_file_link_tag}');"
```
Comment (GBH) [inline app/views/comfy/admin/cms/fragments/_form_fragment_attachments.html.haml]: Don't think we can do inline JS like that if CSP is enabled. Let's move that into `data` attribute. Also how would one even know that this is something draggable?

File: app/views/comfy/admin/cms/fragments/_form_fragment_attachments.html.haml [hunk 2]
```diff
@@ -4,9 +4,19 @@
...
-      = link_to truncated_filename, attachment, data: {toggle: "page-file-popover", content: thumb}, class: "btn btn-light text-truncate", target: "_blank"
+      :ruby
+        filename = attachment.filename.to_s
+        truncated_filename = truncate(filename, length: 40, omission: "...#{filename.last(10)}")
+        cms_page_file_link_tag = [
+          "{{ cms:page_file_link #{fragment_id}",
+          (", filename: \"#{filename}\"" if multiple),
+          (", as: image" if attachment.image?),
+          " }}".html_safe
+        ].compact.join
```
Comment (GBH) [inline app/views/comfy/admin/cms/fragments/_form_fragment_attachments.html.haml]: Kinda ugly. Let's move that into admin view helper.

File: lib/comfortable_mexican_sofa/content/tags/file.rb [hunk 3]
```diff
@@ -1,5 +1,7 @@
 # frozen_string_literal: true
 
+require "comfortable_mexican_sofa/content/tags/file_content.rb"
```
Comment (GBH) [inline lib/comfortable_mexican_sofa/content/tags/file.rb]: require_relative ?

File: lib/comfortable_mexican_sofa/content/tags/file_link.rb [hunk 4]
```diff
@@ -1,5 +1,7 @@
 # frozen_string_literal: true
 
+require "comfortable_mexican_sofa/content/tags/file_content.rb"
```
Comment (GBH) [inline lib/comfortable_mexican_sofa/content/tags/file_link.rb]: require_relative

File: lib/comfortable_mexican_sofa/content/tags/page_file_link.rb [hunk 5]
```diff
@@ -0,0 +1,80 @@
+# frozen_string_literal: true
+
+require "comfortable_mexican_sofa/content/tags/file_content.rb"
```
Comment (GBH) [inline lib/comfortable_mexican_sofa/content/tags/page_file_link.rb]: require_relative

File: lib/comfortable_mexican_sofa/content/tags/page_file_link.rb [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+  # @return [ActiveStorage::Blob]
+  def file
+    @file ||= if fragment.nil?
+                nil
+              elsif filename.nil?
+                fragment.attachments.first
+              else
+                fragment.attachments.detect { |a| a.filename.to_s == filename }
+              end
```
Comment (GBH) [inline lib/comfortable_mexican_sofa/content/tags/page_file_link.rb]: Don't like massive offsets like these. Should be:
```
foo ||=
  if bar?
```

I thought rubocop would catch this

File: app/views/comfy/admin/cms/fragments/_form_fragment_attachments.html.haml (hunk 1 above)
Comment (glebm) [reply app/views/comfy/admin/cms/fragments/_form_fragment_attachments.html.haml]: Done. I didn't write in CoffeeScript, but it's only a few line so perhaps you could rewrite it if you really want it to be CoffeeScript?

> Also how would one even know that this is something draggable?

True that it's not very discoverable, but it does have `cursor: pointer` on hover, and I've now made a thumbnail display for images while dragged.

File: app/views/comfy/admin/cms/fragments/_form_fragment_attachments.html.haml (hunk 2 above)
Comment (glebm) [reply app/views/comfy/admin/cms/fragments/_form_fragment_attachments.html.haml]: Done

File: lib/comfortable_mexican_sofa/content/tags/file.rb (hunk 3 above)
Comment (glebm) [reply lib/comfortable_mexican_sofa/content/tags/file.rb]: Done

File: lib/comfortable_mexican_sofa/content/tags/page_file_link.rb (hunk 5 above)
Comment (glebm) [reply lib/comfortable_mexican_sofa/content/tags/page_file_link.rb]: Done

File: lib/comfortable_mexican_sofa/content/tags/page_file_link.rb (hunk 6 above)
Comment (glebm) [reply lib/comfortable_mexican_sofa/content/tags/page_file_link.rb]: Done

**Comment by glebm** [comment]
I'll also move page file popovers into the same file

**Comment by glebm** [comment]
All done

**Comment by glebm** [comment]
There is a small glitch with the popover re-appearing at the end of the drop:

gif

Looking into it

**Comment by glebm** [comment]
Firefox glitch fixed

**Comment by glebm** [comment]
Files can now also be dragged out of the global files modal

**Comment by glebm** [comment]
OK, I'll try. Everything else looks good to you?

**Comment by glebm** [comment]
I wrote a system test as far as I could take it, because `drag_to` does not seem to work! Tried with the latest chromedriver too.

**Comment by glebm** [comment]
I've now written tests for both global files and page files drag'n'drop but they don't work and so are currently being skipped.

I've tried for a while to fix them but I couldn't. I won't look into this further.

**Comment by glebm** [comment]
Cool kids these days use TypeScript

**Comment by glebm** [comment]
I've removed the extra fixtures.

> I'm going to take a look at the drag_to issue.

Start with the page files test as it's simpler (no iframes).

**Comment by glebm** [comment]
I've made a video for the release notes with [Peek](https://github.com/phw/peek):

cms file drag n drop

**Comment by GBH** [comment]
I'm in process of merging. Running locally and it's great. Dragging files from file pop-up to places is awesome. Gonna leave comments, but just as notes.

File: test/test_helper.rb [hunk 7]
```diff
@@ -174,6 +174,18 @@ def read_file(filename)
...
+  /usr/bin/chromium-browser
+  /Applications/Chromium.app/Contents/MacOS/Chromium
+  /Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome
+].find { |path| File.executable?(path) }
+Selenium::WebDriver::Chrome.driver_path = ENV["CHROMEDRIVER_PATH"] || %w[
+  /usr/bin/chromedriver
+  /usr/lib/chromium-browser/chromedriver
+  /usr/local/bin/chromedriver
+].find { |path| File.executable?(path) }
+
```
Comment (GBH) [inline test/test_helper.rb]: Had to remove this. Broke it for me as it wouldn't find my chromedriver that lives in a different bin directory (but still on my $PATH). There has to be a documentation on Rails Guides for installing all the crap. Guides only have a bit about testing, but nothing about how to get it all running in the first place.

File: lib/comfortable_mexican_sofa/content/tags/file_content.rb [hunk 8]
```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+# A mixin for tags that returns the file as their content.
+module ComfortableMexicanSofa::Content::Tag::FileContent
```
Comment (GBH) [inline lib/comfortable_mexican_sofa/content/tags/file_content.rb]: moving this file into /mixins as it's not a tag

File: app/views/comfy/admin/cms/files/_file.html.haml [hunk 9]
```diff
@@ -1,4 +1,7 @@
 %li{data: {id: file.id}}
+  :ruby
+    file_tag = "{{ cms:file_link #{file.id} }}"
```
Comment (GBH) [inline app/views/comfy/admin/cms/files/_file.html.haml]: I think I want the same dealio to generate `as: image` for image attachments. Will expand that helper (or make new one) to do this.

File: test/system/files_frontend_test.rb [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+require_relative "../test_helper"
+
+class FilesFrontendTest < ApplicationSystemTestCase
+
+  setup do
+    @site = comfy_cms_sites(:default)
+  end
+
+  def test_site_file_drag_and_drop
```
Comment (GBH) [inline test/system/files_frontend_test.rb]: seems like we can't test this due to chromedriver limitation. leaving skip but removing code.

File: test/system/files_frontend_test.rb [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    TEXT
+    visit_p edit_comfy_admin_cms_site_page_path(site_id: @site, id: comfy_cms_pages(:default))
+    attach_file Rails.root.join("test/fixtures/files/image.jpg"),
+                name: "page[fragments_attributes][1][files][]", multiple: true
+    click_button I18n.t("comfy.admin.cms.pages.form.update")
+    assert_link "image.jpg"
+
+    # Dragging the link should generate CMS markup.
+    skip "Drag'n'drop does not work for some reason"
+    find_link("image.jpg").drag_to(content_field)
```
Comment (GBH) [inline test/system/files_frontend_test.rb]: Can't figure out this one either. By all accounts it should work. Dragging other links should work too (to eliminate js fuckery) but it does not. Maybe warrants filing issue with Capybara.

**Comment by GBH** [comment]
@glebm can you double check if everything works in Firefox? On my VM (Firefox 50 though) none of the draggable stuff works as in Chrome. Dragging from iframe doesn't work at all and dragging from page files dumps link's url and not the tag's string.

I'm trying to upgrade to latest Firefox here. If latest version works then no worries.

**Comment by glebm** [comment]
It works for me perfectly in Firefox 58 on Ubuntu 16.04. Firefox 50 is from June 2016.

**Comment by GBH** [comment]
Got Firefox 58 on LinuxMint (Ubuntu). Page file link drag works perfectly. One thing that still doesn't work is that drag from the iframe. In Chrome iframe goes away and it's possible to drag into form input. In Firefox iframe sticks around.

**Comment by glebm** [comment]
@GBH Reproduced, fixed in #802

**Pull request merged**